File: 3086/main.go
package main

import (
	"fmt"
	"math"
)

func minimumMoves(nums []int, k int, maxChanges int) int64 {
	n := len(nums)
	f := func(i int) int {
		x := nums[i]
		if i - 1 >= 0 {
			x += nums[i - 1]
		}
		if i + 1 < n {
			x += nums[i + 1]
		}
		return x
	}

	indexSum, sum := make([]int64, n + 1), make([]int64, n + 1)
	for i := 0; i < n; i++ {
		indexSum[i + 1] = indexSum[i] + int64(nums[i]) * int64(i)
		sum[i + 1] = sum[i] + int64(nums[i])
	}
	var res int64 = math.MaxInt64
	for i := 0; i < n; i++ {
		if f(i) + maxChanges >= k {
			if k <= f(i) {
				res = min(res, int64(k - nums[i]))
			} else {
				res = min(res, int64(2 * k - f(i) - nums[i]))
			}
			continue
		}
		left, right := 0, n
		for left <= right {
			mid := (left + right) / 2
			i1, i2 := max(i - mid, 0), min(i + mid, n - 1)
			if sum[i2 + 1] - sum[i1] >= int64(k - maxChanges) {
				right = mid - 1
			} else {
				left = mid + 1
			}
		}
		i1, i2 := max(i - left, 0), min(i + left, n - 1)
		if sum[i2 + 1] - sum[i1] > int64(k - maxChanges) {
			i1++
		}
		count1, count2 := sum[i + 1] - sum[i1], sum[i2 + 1] - sum[i + 1]
		res = min(res, indexSum[i2 + 1] - indexSum[i + 1] - int64(i) * count2 + int64(i) * count1 - (indexSum[i + 1] - indexSum[i1]) + 2 * int64(maxChanges))
	}
	return res
}

func main() {
	fmt.Println(minimumMoves([]int{0,0,0,0}, 2, 3))
}

File: 3086/main.rs
struct Solution {}

impl Solution {
    pub fn minimum_moves(nums: Vec<i32>, k: i32, max_changes: i32) -> i64 {
        let n = nums.len() as i32;
        let mut index_sum = vec![0 as i64; (n + 1) as usize];
        let mut sum = vec![0 as i64; (n + 1) as usize];
        for i in 0..n {
            index_sum[(i + 1) as usize] = index_sum[i as usize] + (nums[i as usize] as i64) * (i as i64);
            sum[(i + 1) as usize] = sum[i as usize] + (nums[i as usize] as i64);
        }
        let mut res = (((1 as u64) << 63) - 1) as i64;
        for i in 0..n {
            if Self::f(i, &nums) + max_changes >= k {
                if k <= Self::f(i, &nums) {
                    res = std::cmp::min(res, (k - nums[i as usize]) as i64);
                } else {
                    res = std::cmp::min(res, (2 * k - Self::f(i, &nums) - nums[i as usize]) as i64);
                }
                continue;
            }
            let (mut left, mut right) = (0, n);
            while left <= right {
                let mid = (left + right) / 2;
                let (i1, i2) = (std::cmp::max(i - mid, 0), std::cmp::min(i+mid, n-1));
                if sum[(i2 + 1) as usize] - sum[i1 as usize] >= (k - max_changes) as i64 {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            let (mut i1, i2) = (std::cmp::max(i-left, 0), std::cmp::min(i + left, n - 1));
            if sum[(i2 + 1) as usize] - sum[i1 as usize] > (k - max_changes) as i64 {
                i1 += 1;
            }
            let (count1, count2) = (sum[(i + 1) as usize] - sum[i1 as usize], sum[(i2 + 1) as usize] - sum[(i + 1) as usize]);
            res = std::cmp::min(res, index_sum[(i2 + 1) as usize] - index_sum[(i + 1) as usize] - (i as i64) * count2 + (i as i64) * count1- (index_sum[(i + 1) as usize] - index_sum[i1 as usize]) + (2 * max_changes as i64));
        }
        res
    }

    fn f(i: i32, nums: &Vec<i32>) -> i32 {
        let mut x = nums[i as usize];
        if i - 1 >= 0 {
            x += nums[(i - 1) as usize];
        }
        if (i + 1) < nums.len() as i32 {
            x += nums[(i + 1) as usize];
        }
        x
    }
}

fn main() {
    println!("{}", Solution::minimum_moves(vec![0,0,0,0], 2, 3))
}